File: server/services/peptideForumService.js
const { peptideForumStore } = require('../storage');
const mysqlClient = require('../database/mysqlClient');
const { logger } = require('../config/logger');

const normalizeText = (value) => {
  const text = typeof value === 'string' ? value.trim() : '';
  return text || '';
};

const normalizeOptionalText = (value) => {
  const text = normalizeText(value);
  return text ? text : null;
};

const normalizeDurationMinutes = (item) => {
  const raw =
    item?.durationMinutes ??
    item?.duration_minutes ??
    item?.duration ??
    item?.lengthMinutes ??
    item?.length_minutes ??
    null;
  const value = typeof raw === 'number' ? raw : Number(normalizeText(raw));
  return Number.isFinite(value) && value > 0 ? Math.round(value) : null;
};

const tryParseDateTime = (dateValue, timeValue) => {
  const rawDate = normalizeText(dateValue);
  const rawTime = normalizeText(timeValue);
  if (!rawDate) return { iso: null, rawDate: null, rawTime: rawTime || null };

  const candidates = [];
  if (rawTime) {
    // Google Sheet time strings like "3:00 PM" should be interpreted in PST.
    candidates.push(`${rawDate} ${rawTime} GMT-0800`);
    candidates.push(`${rawDate} ${rawTime}`);
  }
  candidates.push(rawDate);

  for (const candidate of candidates) {
    const parsed = new Date(candidate);
    if (!Number.isNaN(parsed.getTime())) {
      return { iso: parsed.toISOString(), rawDate, rawTime: rawTime || null };
    }
  }

  return { iso: null, rawDate, rawTime: rawTime || null };
};

const isCalendarDateOnly = (value) =>
  /^\d{4}-\d{1,2}-\d{1,2}$/.test(String(value || '').trim()) ||
  /^\d{1,2}\/\d{1,2}\/\d{2,4}$/.test(String(value || '').trim());

const tryParseEndDateTime = (dateValue, timeValue) => {
  const rawDate = normalizeText(dateValue);
  const rawTime = normalizeText(timeValue);
  if (rawDate && !rawTime && isCalendarDateOnly(rawDate)) {
    return tryParseDateTime(rawDate, '23:59:59');
  }
  return tryParseDateTime(dateValue, timeValue);
};

const isLikelyUrl = (value) => /^https?:\/\//i.test(String(value || '').trim());

const normalizeItem = (item, index) => {
  const title = normalizeText(item?.title);
  const description = normalizeOptionalText(item?.description);
  const link = normalizeText(item?.link);
  const recording = normalizeOptionalText(item?.recording);
  const { iso: dateIso, rawDate, rawTime } = tryParseDateTime(item?.date, item?.time);
  const endDateInput =
    item?.endDate ?? item?.end_date ?? item?.endsAt ?? item?.ends_at ?? item?.endAt ?? item?.end_at ?? item?.end;
  const endTimeInput = item?.endTime ?? item?.end_time ?? item?.endTimeRaw ?? item?.end_time_raw;
  const { iso: endDateIso, rawDate: endRawDate, rawTime: endRawTime } = tryParseEndDateTime(
    endDateInput,
    endTimeInput,
  );
  const durationMinutes = normalizeDurationMinutes(item);

  if (!title && !link) {
    return { ok: false, error: `Row ${index}: missing title and link` };
  }
  if (link && !isLikelyUrl(link)) {
    return { ok: false, error: `Row ${index}: invalid link` };
  }
  if (recording && !isLikelyUrl(recording)) {
    return { ok: false, error: `Row ${index}: invalid recording link` };
  }

  const idBase = `${title || 'class'}|${dateIso || rawDate || 'nodate'}|${rawTime || 'notime'}|${link || 'nolink'}|${recording || 'norecording'}`;
  const id = Buffer.from(idBase).toString('base64url').slice(0, 48);

  return {
    ok: true,
    value: {
      id,
      title: title || (link ? 'The Peptide Forum' : 'Untitled'),
      date: dateIso || rawDate || null,
      time: rawTime || null,
      description,
      link: link || null,
      recording,
      endDate: endDateIso || endRawDate || null,
      endDateRaw: endRawDate || null,
      endTime: endRawTime || null,
      durationMinutes,
    },
  };
};

const list = () => {
  const payload = peptideForumStore.read();
  const items = Array.isArray(payload?.items) ? payload.items : [];
  const updatedAt = typeof payload?.updatedAt === 'string' ? payload.updatedAt : null;
  return { updatedAt, items };
};

const toSqlDateTime = (value) => {
  if (!value) return null;
  const parsed = value instanceof Date ? value : new Date(value);
  if (Number.isNaN(parsed.getTime())) return null;
  return parsed.toISOString().slice(0, 19).replace('T', ' ');
};

const persistToMysql = async (items) => {
  if (!mysqlClient.isEnabled()) {
    return {
      mysqlEnabled: false,
      stored: 0,
      removed: 0,
    };
  }

  const syncToken = `${Date.now().toString(36)}${Math.random().toString(36).slice(2, 8)}`;
  let stored = 0;
  for (const item of items) {
    await mysqlClient.execute(
      `
        INSERT INTO peptide_forum_items (
          id,
          title,
          event_date,
          event_date_raw,
          event_time_raw,
          event_end_date,
          event_end_date_raw,
          event_end_time_raw,
          duration_minutes,
          description,
          link,
          recording,
          sync_token
        ) VALUES (
          :id,
          :title,
          :eventDate,
          :eventDateRaw,
          :eventTimeRaw,
          :eventEndDate,
          :eventEndDateRaw,
          :eventEndTimeRaw,
          :durationMinutes,
          :description,
          :link,
          :recording,
          :syncToken
        )
        ON DUPLICATE KEY UPDATE
          title = VALUES(title),
          event_date = VALUES(event_date),
          event_date_raw = VALUES(event_date_raw),
          event_time_raw = VALUES(event_time_raw),
          event_end_date = VALUES(event_end_date),
          event_end_date_raw = VALUES(event_end_date_raw),
          event_end_time_raw = VALUES(event_end_time_raw),
          duration_minutes = VALUES(duration_minutes),
          description = VALUES(description),
          link = VALUES(link),
          recording = VALUES(recording),
          sync_token = VALUES(sync_token)
      `,
      {
        id: item.id,
        title: item.title,
        eventDate: toSqlDateTime(item.date),
        eventDateRaw: item.date ? String(item.date) : null,
        eventTimeRaw: item.time ? String(item.time) : null,
        eventEndDate: toSqlDateTime(item.endDate),
        eventEndDateRaw: item.endDateRaw ? String(item.endDateRaw) : null,
        eventEndTimeRaw: item.endTime ? String(item.endTime) : null,
        durationMinutes:
          typeof item.durationMinutes === 'number' && Number.isFinite(item.durationMinutes)
            ? Math.round(item.durationMinutes)
            : null,
        description: item.description || null,
        link: item.link || null,
        recording: item.recording || null,
        syncToken,
      },
    );
    stored += 1;
  }

  let removed = 0;
  if (items.length === 0) {
    const deletedAll = await mysqlClient.execute('DELETE FROM peptide_forum_items');
    removed = Number(deletedAll?.affectedRows || 0);
  } else {
    const deletedStale = await mysqlClient.execute(
      'DELETE FROM peptide_forum_items WHERE sync_token <> :syncToken',
      { syncToken },
    );
    removed = Number(deletedStale?.affectedRows || 0);
  }

  return {
    mysqlEnabled: true,
    stored,
    removed,
  };
};

const replaceFromWebhook = async (incoming) => {
  const rows = Array.isArray(incoming) ? incoming : [];
  const errors = [];
  const items = [];

  rows.forEach((row, idx) => {
    const normalized = normalizeItem(row, idx);
    if (!normalized.ok) {
      errors.push(normalized.error);
      return;
    }
    items.push(normalized.value);
  });

  const next = {
    updatedAt: new Date().toISOString(),
    items,
  };
  peptideForumStore.write(next);

  let mysql = null;
  try {
    mysql = await persistToMysql(items);
  } catch (error) {
    logger.error({ err: error }, '[Peptide Forum] Failed to persist forum rows to MySQL');
    mysql = {
      mysqlEnabled: mysqlClient.isEnabled(),
      stored: 0,
      removed: 0,
      error:
        typeof error?.message === 'string' && error.message
          ? error.message
          : 'Unable to persist forum rows to MySQL',
    };
  }

  return {
    updatedAt: next.updatedAt,
    stored: items.length,
    received: rows.length,
    errors,
    mysql,
  };
};

module.exports = {
  list,
  replaceFromWebhook,
};

File: python_backend/services/peptide_forum_service.py
from __future__ import annotations

import base64
import logging
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Tuple

from ..repositories import peptide_forum_repository
from ..services import get_config
from ..storage import peptide_forum_store

logger = logging.getLogger(__name__)


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def _to_str(value: Any) -> str:
    return "" if value is None else str(value)


def _norm_text(value: Any) -> str:
    return _to_str(value).strip()


def _try_parse_date(value: Any) -> Tuple[Optional[str], Optional[str]]:
    raw = _norm_text(value)
    if not raw:
        return None, None
    try:
        parsed = datetime.fromisoformat(raw.replace("Z", "+00:00"))
        if parsed.tzinfo is None:
            parsed = parsed.replace(tzinfo=timezone.utc)
        return parsed.astimezone(timezone.utc).isoformat().replace("+00:00", "Z"), raw
    except Exception:
        return None, raw


def list_items() -> Dict[str, Any]:
    config = get_config()
    if bool(config.mysql.get("enabled")):
        items = peptide_forum_repository.list_posts(limit=500)
        return {"updatedAt": _now_iso(), "items": items}

    if not peptide_forum_store:
        return {"updatedAt": None, "items": []}
    payload = peptide_forum_store.read() or {}
    updated_at = payload.get("updatedAt") if isinstance(payload.get("updatedAt"), str) else None
    items = payload.get("items") if isinstance(payload.get("items"), list) else []
    return {"updatedAt": updated_at, "items": items}


def _normalize_item(item: Any, index: int) -> Tuple[Optional[Dict[str, Any]], Optional[str]]:
    title = _norm_text((item or {}).get("title") if isinstance(item, dict) else "")
    description = _norm_text((item or {}).get("description") if isinstance(item, dict) else "")
    link = _norm_text((item or {}).get("link") if isinstance(item, dict) else "")
    date_iso, date_raw = _try_parse_date((item or {}).get("date") if isinstance(item, dict) else None)

    if not title and not link:
        return None, f"Row {index}: missing title and link"

    id_base = f"{title or 'post'}|{date_iso or date_raw or 'nodate'}|{link or 'nolink'}"
    post_id = base64.urlsafe_b64encode(id_base.encode("utf-8")).decode("utf-8").rstrip("=")[:48]

    return (
        {
            "id": post_id,
            "title": title or ("The Peptide Forum" if link else "Untitled"),
            "date": date_iso or date_raw or None,
            "description": description or None,
            "link": link or None,
        },
        None,
    )


def replace_from_webhook(items: List[Dict[str, Any]]) -> Dict[str, Any]:
    normalized_items: List[Dict[str, Any]] = []
    errors: List[str] = []

    rows = items if isinstance(items, list) else []
    for idx, row in enumerate(rows):
        value, error = _normalize_item(row, idx)
        if error:
            errors.append(error)
            continue
        if value:
            normalized_items.append(value)

    config = get_config()
    updated_at = _now_iso()

    if bool(config.mysql.get("enabled")):
        for post in normalized_items:
            date_value = post.get("date")
            date_iso, date_raw = _try_parse_date(date_value)
            date_at = None
            if date_iso:
                try:
                    date_at = (
                        datetime.fromisoformat(date_iso.replace("Z", "+00:00"))
                        .astimezone(timezone.utc)
                        .replace(tzinfo=None)
                    )
                except Exception:
                    date_at = None

            peptide_forum_repository.upsert_post(
                {
                    "id": post.get("id"),
                    "title": post.get("title"),
                    "date_at": date_at,
                    "date_raw": date_raw,
                    "description": post.get("description"),
                    "link": post.get("link"),
                }
            )

        # Treat webhook payload as authoritative (mirror the sheet).
        peptide_forum_repository.delete_missing_ids(
            [p.get("id") for p in normalized_items if p.get("id")]
        )
    else:
        payload = {"updatedAt": updated_at, "items": normalized_items}
        if peptide_forum_store:
            peptide_forum_store.write(payload)
        else:
            logger.warning("peptide_forum_store not initialized; skipping persistence")

    return {
        "updatedAt": updated_at,
        "received": len(rows),
        "stored": len(normalized_items),
        "errors": errors,
    }
